govcd/edgegateway_test.go

```diff
@@ -743,31 +743,24 @@ func testCheckFirewallConfig(beforeFw *types.FirewallConfigWithXml, beforeFwXml
 	afterFw.Version = ""
 
 	// Because the test enables and disables firewall configuration, main firewall rule has its ID
-	// and ruleTag changed during the test and we must ignore this change while comparing. This is
-	// always the first rule kept at the top so we are replacing `id` and `ruletag` fields in
-	// "before" and "after" values.
-	beforeFw.FirewallRules.Text = replaceFirstMatch(beforeFw.FirewallRules.Text, `<id>\d*</id>`, "<id>99999</id>")
-	beforeFw.FirewallRules.Text = replaceFirstMatch(beforeFw.FirewallRules.Text, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
-	afterFw.FirewallRules.Text = replaceFirstMatch(afterFw.FirewallRules.Text, `<id>\d*</id>`, "<id>99999</id>")
-	afterFw.FirewallRules.Text = replaceFirstMatch(afterFw.FirewallRules.Text, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
-
-	beforeFwXml = replaceFirstMatch(beforeFwXml, `<id>\d*</id>`, "<id>99999</id>")
-	beforeFwXml = replaceFirstMatch(beforeFwXml, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
-	afterFwXml = replaceFirstMatch(afterFwXml, `<id>\d*</id>`, "<id>99999</id>")
-	afterFwXml = replaceFirstMatch(afterFwXml, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
+	// and ruleTag changed during the test and we must ignore this change while comparing.
+	beforeFw.FirewallRules.Text = replaceAllMatches(beforeFw.FirewallRules.Text, `<id>\d*</id>`, "<id>99999</id>")
+	beforeFw.FirewallRules.Text = replaceAllMatches(beforeFw.FirewallRules.Text, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
+	afterFw.FirewallRules.Text = replaceAllMatches(afterFw.FirewallRules.Text, `<id>\d*</id>`, "<id>99999</id>")
+	afterFw.FirewallRules.Text = replaceAllMatches(afterFw.FirewallRules.Text, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
+
+	beforeFwXml = replaceAllMatches(beforeFwXml, `<id>\d*</id>`, "<id>99999</id>")
+	beforeFwXml = replaceAllMatches(beforeFwXml, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
+	afterFwXml = replaceAllMatches(afterFwXml, `<id>\d*</id>`, "<id>99999</id>")
+	afterFwXml = replaceAllMatches(afterFwXml, `<ruleTag>\d*</ruleTag>`, "<ruleTag>99999</ruleTag>")
```

Comment:
Curious, what changed in API that you had to switch from `replaceFirstMatch` to `replaceAllMatches`? Thanks!

Reply:
Interested as well. The APIs behind it should be the proxied ones so I would expect they don't depend on versions of vCD API

Reply:
Well you assumptions wasn't right. It not always return first. Sometimes - not. Usually it works, but sometimes failed.

Reply:
Is there any way you know  how to replicate this as I've never hit it (at least with v27 )? I can look at it. I am exactly afraid of replacing all `id`s because this may compromise the goal of this test.

Reply:
I'd like more info on this change, as it alters the test behavior substantially. What exactly is failing that warrants such a modification?

Reply:
Test lack explicit value checking and expects that vCD always returns XML list elements to same order. So just check savedXML against newXML - to verify that everything is same. In general is the same, but sometimes vCD returns list in different order and XML don't match, though values of xml is the same.

Reply:
If XML rule order changes - then it means that real rule processing order is changed during global firewall setting changed, which is bad. I need to catch it.

Is it the `TestEdgeGateway_UpdateFwGeneralParams` test run? I will try to leave it running overnight on v29 and see if it fails.

Reply:
TestEdgeGateway_UpdateFwGeneralParams is enough.

Reply:
I have ran it 3812 times over night on v29 and it didn't fail. I suspect this is because your edge gateway firewall rule list is not empty for testing (looks like it has more than 1 rule of type `internal_high`). That isn't perfect, but I'd prefer to split this issue to a separate one (I can pick it on me).

Reply:
We have agreed to split this issue out of version change PR to minimize its footprint and I have raised a separate issue for investigation of this problem #264

Reply:
As we agree - revert back